File: Python/beginner/04_GraphSearch/boj_24445.py
import sys
from collections import deque

input = sys.stdin.readline

def solve(grp: list[list[int]], start: int) -> None:
    visited = [False] * len(grp)
    visited[start] = True
    q = deque([start])

    ans = [0] * (len(grp))
    cnt = 1

    while q:
        now = q.popleft()
        ans[now] = cnt
        cnt += 1

        for nst in grp[now]:
            if not visited[nst]:
                visited[nst] = True
                q.append(nst)

    for i in range(1, len(ans)):
        print(ans[i])


if __name__ == "__main__":
    n, m, r = map(int, input().split())
    grp = [[] for _ in range(n + 1)]
    for _ in range(m):
        u, v = map(int, input().split())
        grp[u].append(v)
        grp[v].append(u)

    for g in grp:
        g.sort(reverse=True)

    solve(grp, r)

        

File: Cpp/Beginner/04_GraphSearch/boj_24445.cpp
//
// Created by seungsu on 22. 11. 23.
//
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

void bfs(vector<vector<int>>& grp, int start) {
    queue<int> q;
    vector<bool> visited(grp.size(), false);

    q.push(start);
    visited[start] = true;

    vector<int> rst(grp.size(), 0);
    int cnt = 1;

    while (!q.empty()) {
        int now = q.front();
        q.pop();

        rst[now] = cnt++;

        for (int k : grp[now]) {
            if (!visited[k]) {
                q.push(k);
                visited[k] = true;
            }
        }
    }

    for (int i = 1; i < rst.size(); i++)
        cout << rst[i] << "\n";
}

void boj_24445() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, m, r;
    cin >> n >> m >> r;

    vector<vector<int>> grp(n + 1, vector<int>());
    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;

        grp[a].push_back(b);
        grp[b].push_back(a);
    }

    for (int i = 0; i < grp.size(); i++) {
        sort(grp[i].begin(), grp[i].end(), [](int a, int b) {
            return a > b;
        });
    }

    bfs(grp, r);
}